ctapipe/reco/shower_reconstructor.py

```diff
@@ -0,0 +1,208 @@
+from functools import partial
+
+import astropy.units as u
+import numpy as np
+from astropy.coordinates import AltAz, SkyCoord
+from iminuit import Minuit
+
+from ctapipe.containers import Model3DReconstructedGeometryContainer
+from ctapipe.coordinates import CameraFrame, TelescopeFrame
+from ctapipe.core.traits import Unicode
+from ctapipe.image import (
+    GaussianShowermodel,
+    ShowermodelPredictor,
+    neg_log_likelihood_approx,
+)
+from ctapipe.reco.reco_algorithms import Reconstructor
+
+
+class Model3DGeometryReconstructor(Reconstructor):
+    geometry_seed = Unicode(default_value="HillasReconstructor").tag(config=True)
+
+    def __init__(self, subarray, **kwargs):
+        super().__init__(subarray=subarray, **kwargs)
+
+    def __call__(self, event):
+        if self.geometry_seed not in event.dl2.stereo.geometry:
+            raise ValueError()
+
+        # get all telescope properties that we need for the ShowermodelPredictor
+        tel_positions = {}
```

Comment:
As said above, give the subarray to the `Predictor` and move that to `__init__`. It doesn't change so shouldn't be called in the event loop.